pkg/binary/envoy/run_test.go

```diff
@@ -32,98 +32,111 @@ import (
+			args:      []string{"quiet_exit=3"},
+			expectedStdout: fmt.Sprintf(`starting: %s quiet_exit=3
+working directory: %s
+`, fakeEnvoy, workingDir),
+			expectedStderr: "started\n",
+			expectedErr:    "envoy exited with status: 3",
+			expectedHooks:  []string{"preStart"},
 		},
 	}
 
 	for _, tt := range tests {
 		tc := tt
 		t.Run(tc.name, func(t *testing.T) {
-			debugDir, removeDebugDir := morerequire.RequireNewTempDir(t)
-			defer removeDebugDir()
-
-			fakeEnvoy, removeFakeEnvoy := morerequire.RequireCaptureScript(t, "envoy")
-			defer removeFakeEnvoy()
-
-			fakeTimestamp := "1619574747231823000"
-			workingDir := filepath.Join(debugDir, fakeTimestamp)
-			fakeLogger := log.New(io.Discard, "", 0)
-			o := &globals.RunOpts{EnvoyPath: fakeEnvoy, WorkingDir: workingDir, Log: fakeLogger, DebugLog: fakeLogger}
+			o := &globals.RunOpts{EnvoyPath: fakeEnvoy, WorkingDir: workingDir}
 			require.NoError(t, os.MkdirAll(workingDir, 0750))
 
-			// This ensures functions are called in the correct order
-			r, preStartCalled, preTerminationCalled, postTerminationCalled := newRuntimeWithMockFunctions(t, o)
+			stdout := new(bytes.Buffer)
+			stderr := new(bytes.Buffer)
+			r, hooksCalled := newRuntimeWithMockHooks(t, stdout, stderr, o)
+
+			err := envoytest.RequireRunTerminate(t, tc.terminate, r, tc.args...)
+			if tc.expectedErr == "" {
+				require.NoError(t, err)
+			} else {
+				require.EqualError(t, err, tc.expectedErr)
+			}
 
-			envoytest.RequireRunTerminate(t, tc.terminate, r, tc.args...)
+			// Assert appropriate hooks are called
+			require.Equal(t, tc.expectedHooks, *hooksCalled)
 
-			// Assert appropriate functions are called
-			require.True(t, *preStartCalled, "preStart was not called")
-			require.Equal(t, tc.wantPreTerm, *preTerminationCalled, "expected preTermination execution to be %v", tc.wantPreTerm)
-			require.Equal(t, tc.wantPostTerm, *postTerminationCalled, "expected postTermination execution to be %v", tc.wantPostTerm)
+			// Validate we ran what we thought we did
+			require.Equal(t, tc.expectedStdout, stdout.String())
+			require.Equal(t, tc.expectedStderr, stderr.String())
 
-			// Ensure the debug directory was created
+			// Ensure the working directory was deleted, and the debug directory only contains the archive
 			files, err := os.ReadDir(debugDir)
 			require.NoError(t, err)
 			require.Equal(t, 1, len(files))
+			archive := filepath.Join(debugDir, files[0].Name())
+			require.Equal(t, filepath.Join(debugDir, fakeTimestamp+".tar.gz"), archive)
 
-			// Ensure a debug archive was created
-			require.Contains(t, files[0].Name(), ".tar.gz")
+			// Cleanup for the next run
+			require.NoError(t, os.Remove(archive))
 		})
 	}
 }
 
 // This ensures functions are called in the correct order
-func newRuntimeWithMockFunctions(t *testing.T, o *globals.RunOpts) (*envoy.Runtime, *bool, *bool, *bool) {
+func newRuntimeWithMockHooks(t *testing.T, stdout, stderr io.Writer, o *globals.RunOpts) (*envoy.Runtime, *[]string) {
```

Comment:
I don't think this should be pointer (slice is itself is basically a pointer (plus len and cap).
```suggestion
func newRuntimeWithMockHooks(t *testing.T, stdout, stderr io.Writer, o *globals.RunOpts) (*envoy.Runtime, []string) {
```

Reply:
heh didn't work :)

Reply:
it copies the slice, so basically the updates aren't visible